indico/modules/events/models/persons.py

```diff
@@ -148,6 +149,69 @@ def for_user(cls, user, event):
         person = event.persons.filter_by(user=user).first()
         return person or cls.create_from_user(user, event)
 
+    @classmethod
+    def link_user_by_email(cls, user):
+        """
+        Links all email-based persons matching the user's
+        email addresses with the user.
+
+        :param user: A User object.
+        """
+        from indico.modules.events.models.events import Event
+        query = (cls.query
+                 .options(noload('*'))
+                 .join(EventPerson.event_new)
+                 .filter(~Event.is_deleted,
+                         cls.email.in_(user.all_emails),
+                         cls.user_id.is_(None)))
+        for event_person in query:
+            existing = (cls.query
+                        .options(noload('*'))
+                        .filter_by(user=user, event_id=event_person.event_id)
+                        .join(EventPerson.user)
+                        .first())
+            if existing is None:
+                event_person.user = user
+            else:
+                existing.merge_person_info(event_person)
+                db.session.delete(event_person)
+        db.session.flush()
+
+    def merge_person_info(self, other):
+        from indico.modules.events.contributions.models.persons import AuthorType
+        for column_name in {'_title', 'affiliation', 'address', 'phone', 'first_name', 'last_name'}:
+            value = getattr(self, column_name) or getattr(other, column_name)
+            setattr(self, column_name, value)
+
+        def _get_author_type(src_link, dest_link):
+            if AuthorType.primary in (src_link.author_type, dest_link.author_type):
+                return AuthorType.primary
+            if src_link.author_type == AuthorType.none:
```

Comment:
if..elif..else would be more readable (all returns on the same indentation level)